spec/acts_as_tag_pgarray/taggable_spec.rb

```diff
@@ -1,162 +1,275 @@
+      end
+
+      it "returns filtered tags for tag_name with prepended scope and block" do
+        expect(User.where("tag like ?", "bl%").all_colors { where(name: ["Ken", "Joe"]) }).to match_array([@user2, @user3].map(&:colors).flatten.uniq.select { |name| name.start_with? "bl" })
+      end
     end
 
-    it "returns filtered tags for tag_name with prepended scope" do
-      expect(User.where("tag like ?", "bl%").all_colors).to match_array([@user1, @user2, @user3].map(&:colors).flatten.uniq.select { |name| name.start_with? "bl" })
-      expect(Admin.where("tag like ?", "bl%").all_colors).to match_array([@admin2].map(&:colors).flatten.uniq.select { |name| name.start_with? "bl" })
+    describe "#colors_cloud" do
+      it "returns tag cloud for tag_name" do
+        expect(User.colors_cloud).to match_array(
+          [@user1, @user2, @user3, @admin1, @admin2].map(&:colors).flatten.group_by(&:to_s).map { |k, v| [k, v.count] }
+        )
+      end
+
+      it "returns filtered tag cloud for tag_name with block" do
+        expect(User.colors_cloud { where(name: ["Ken", "Joe"]) }).to match_array(
+          [@user2, @user3].map(&:colors).flatten.group_by(&:to_s).map { |k, v| [k, v.count] }
+        )
+      end
+
+      it "returns filtered tag cloud for tag_name with prepended scope" do
+        expect(User.where("tag like ?", "bl%").colors_cloud).to match_array(
+          [@user1, @user2, @user3, @admin2].map(&:colors).flatten.group_by(&:to_s).map { |k, v| [k, v.count] }.select { |name, count| name.start_with? "bl" }
+        )
+      end
+
+      it "returns filtered tag cloud for tag_name with prepended scope and block" do
+        expect(User.where("tag like ?", "bl%").colors_cloud { where(name: ["Ken", "Joe"]) }).to match_array(
+          [@user2, @user3].map(&:colors).flatten.group_by(&:to_s).map { |k, v| [k, v.count] }.select { |name, count| name.start_with? "bl" }
+        )
+      end
     end
 
-    it "returns filtered tags for tag_name with prepended scope and bock" do
-      expect(User.where("tag like ?", "bl%").all_colors { where(name: ["Ken", "Joe"]) }).to match_array([@user2, @user3].map(&:colors).flatten.uniq.select { |name| name.start_with? "bl" })
+    describe "with complex scope" do
+      it "works properly" do
+        expect(User.without_any_colors("white").with_any_colors("blue").order(:created_at).limit(10)).to eq [@user1, @user3]
+      end
     end
   end
 
-  describe "#colors_cloud" do
-    it "returns tag cloud for tag_name" do
-      expect(User.colors_cloud).to match_array(
-        [@user1, @user2, @user3, @admin1, @admin2].map(&:colors).flatten.group_by(&:to_s).map { |k, v| [k, v.count] }
-      )
+  context "with allowed option" do
+    context "when options is not an array" do
+      it "raises an error" do
+        expect { User.acts_as_taggable_array_on :colors, allowed: "red, blue" }
```

Comment:
What should happen if allowed is an empty array? Can we test for that as well?

Reply:
Good catch. What do you think about throwing an error in this scenario? Another option would be to allow any value, having the same behavior as when the allowed option is not present.

Reply:
I allowed only has as an empty array, just act like there is no such setting. No need to throw errors.